clojure-mode.el

```diff
@@ -2275,6 +2275,8 @@ This will skip over sexps that don't represent objects, so that ^hints and
   (condition-case nil
       (save-excursion
         (beginning-of-defun)
+        (clojure-forward-logical-sexp 1)
```

Comment:
Perhaps we should add some comments here?

Reply:
The change is to handle whitespace before comment form. Not sure if this is the best way to do it though.
Added a couple of comments now to make the intent more clear.

Reply:
My point was mainly that it looks a bit weird that we're going forward and backwards back-to-back. If that's used often in the code we can make a function I guess.

Reply:
Agreed. But I am not sure what is the correct abstraction anymore as it's more of a hack.

This is the third occurrence in the codebase but there are other places where similar logic was used with different functions and some places where edge cases are not handled. Even my current change will still break if there are multiple expressions in the same line.

Looks like all these extra steps are only needed in many places because of `beginning-of-defun` behaving 'incorrectly'.
For ex:
```
1 (ns abc) (defn abc []
             (comment
               (println "try (clojure-beginning-of-defun-function) from here"))
             "abc")
```
Now running `(clojure-beginning-of-defun-function)` or `(beginning-of-defun)` from any of the inner expressions takes the cursur to the beginning of the line (before 1).
The expected behaviour IMO from the name would be having the cursor before `(defn`.
The docstring mentions this but there is no way to easy way to control the end location of the cursor.

clojure-mode uses beginning-of-defun a lot so there are a lot of unhandled edge cases.
For example take these two code blocks:
```
(defn s1 [x] (str x)) (+ 1 2)

(defn s2
    [x] (str x))
```
```
(defn s1 [x] (str x)) 
(+ 1 2)
(defn s2
    [x] (str x))
```

Run `M-x cider-eval-defun-at-point`(C-c C-c) inside (+ 1 2) and the result is different in the two scenarios.

Reply:
tldr: IIUC bad choice of abstraction from `beginning-of-defun` fn has propagated many edge cases possibly in all lisp modes.
Maybe I am missing something basic or someone already solved this issue.

Reply:
@bbatsov I have raised a separate PR for replacing `beginning-of-defun` fn completely here.
https://github.com/clojure-emacs/clojure-mode/pull/663
After merging that PR these two lines here can be removed.

Reply:
Thanks for diving deep into this!

Reply:
Merged latest master and removed the extra lines here.
Will wait for @vemv to resolve the other conversation.